src/core/util/props.js

```diff
@@ -70,7 +70,8 @@ function getPropDefaultValue (vm: ?Component, prop: PropOptions, key: string): a
     return vm[key]
   }
   // call factory function for non-Function types
-  return typeof def === 'function' && prop.type !== Function
+  // a value is Function if its prototype is function even across different execution context
+  return typeof def === 'function' && (!prop.type || typeof prop.type.prototype !== 'function')
```

Comment:
Actually we already have the proper check implemented here: https://github.com/vuejs/vue/blob/dev/src/core/util/props.js#L166-L169

So we can just do `getType(prop.type) === 'Function'`